oneflow/core/common/data_type.h

```diff
@@ -74,6 +74,16 @@ struct IsFloating : std::integral_constant<bool, false> {};
 OF_PP_FOR_EACH_TUPLE(SPECIALIZE_TRUE_FLOATING, FLOATING_DATA_TYPE_SEQ);
 #undef SPECIALIZE_TRUE_FLOATING
 
+template<>
+struct IsFloating<float16> : std::integral_constant<bool, true> {};
+
+#ifdef WITH_CUDA
+
+template<>
+struct IsFloating<half> : std::integral_constant<bool, true> {};
+
+#endif  // WITH_CUDA
```

Comment:
IsFloating、IsIntegral、... 这些我们自己写的 type traits 有一个问题是没考虑 const 和 volatile，而 std:: is_floating_point 是考虑了的：

> Provides the member constant value which is equal to true, if T is the type float, double, long double, **including any cv-qualified variants**. Otherwise, value is equal to false.

这个 PR 可以顺便修复一下这个问题，可以参考 std::is_floating_point 的实现：

```c++

template< class T >
struct is_floating_point
     : std::integral_constant<
         bool,
         // Note: standard floating-point types
         std::is_same<float, typename std::remove_cv<T>::type>::value
         || std::is_same<double, typename std::remove_cv<T>::type>::value
         || std::is_same<long double, typename std::remove_cv<T>::type>::value
     > {};
```